fhir-server/src/main/java/au/csiro/pathling/fhirpath/element/BooleanPath.java

```diff
@@ -77,4 +78,9 @@ public boolean isComparableTo(@Nonnull final Class<? extends Comparable> type) {
     return COMPARABLE_TYPES.contains(type);
   }
 
+  @Override
+  public boolean canBeCombinedWith(@Nonnull final FhirPath target) {
+    return super.canBeCombinedWith(target) || target instanceof BooleanLiteralPath;
+  }
+
```

Comment:
This really feels somewhat sketchy (for all types) both because of the plethora of special cases and also because the semantics does not match 'mergeWith' in some case, e.g: two StringPaths can always be combined but the merge will fail if they are not of the same fhir type. 

Maybe each FhirPath should have a FhirDefinedType associated with it and then 'canBeCombined` could be in most cases just based on the equality (or equivalence of these types). It may need to be optional to account for UntypedResoucePath. 

A more complex solution would be to allow for implicit type conversions (e.g. Decimal + Integer) or ( URL + String).

Or maybe we should decouple the path hierarchy from the type hierarchy so that all type dependent operation are encapsulated in type classes... e.g. `BooleanType`, `StringType` etc. (including comparison, mateialisation etc),  and then we only have something like `class ElementPath<? extends Type> {  Type type}` which delegates all type related things to it's type object.

Reply:
I don't think that we want coercion as such - I don't think Decimal + Integer should be allowed, but I do think that URL + String should be allowed (as they are the same type in FHIRPath).

I wonder if `mergeWith` is being more strict than it needs to be?

I'll see what I can do to harmonise these concepts within the code.

Reply:
I've made some changes in e990f3caf13759caea449210a2c861cd532c7315 which I hope go some way to addressing your concerns:

* Harmonise the combine/merge functionality used by iif and the combine operator into a single mechanism
* Broaden the set of valid cases to include the combination of paths which have the same schema representation
* Improve test coverage over some of the scenarios that you mentioned